src/incomingDelivery/controller.js

```diff
@@ -292,16 +293,61 @@ function setReceived(req, res) {
     async () => {
       const { _id, receivedQuantities } = req.body;
       const userId = req.user._id;
-
       const incomingDelivery = await IncomingDelivery.findOne({ _id });
+
       if (incomingDelivery.receivedOn)
         return HTTPError("delivery already received");
 
       incomingDelivery.receivedOn = new Date();
       incomingDelivery.receivedBy = userId;
       incomingDelivery.receivedQuantities = receivedQuantities;
-
       await incomingDelivery.save();
+
+      // Joing to packingSlips collection to find qty per item
+      const result = await Shipment.aggregate([
+        {
+          $lookup: {
+            from: PackingSlip.collection.collectionName,
+            localField: "manifest",
+            foreignField: "_id",
+            as: "fromManifest",
+          },
+        },
+        {
+          $match: {
+            _id: incomingDelivery.sourceShipmentId,
+          },
+        },
+      ]);
+
+      // Compare to receivedQuantities. If the  qty is not fullfilled,
+      // make an exact copy
+      result.forEach(async (ogShipment) => {
```

Comment:
In most caess, this should work as expected but in general it's not good to send async callbacks to forEach.
What happens is `myArray.forEach( callback )` will just enqueue a bunch of callbacks and since there is no way to await all those callbacks, your block will just move on to `const data = { ... }; return { data }` and give the client a success message without guarantee that everything executed properly. I.e., this could fail silently and the end-user would never know.

The proper way to synchronize this is something akin to
```
const promises = myArr.map(async (manifest) => ...);
await Promise.all(promises);

return { ... };
```

Reply:
Makes sense. I replaced with maps and promise all rather than awaiting in the foreach.